step4/src/service.js

```diff
@@ -0,0 +1,178 @@
+import {state} from "./state.js";
+import {template} from "./template.js"
+
+function postData(url = '', data = {}) {
+  // Default options are marked with *
+    return fetch(url, {
+        method: 'POST', // *GET, POST, PUT, DELETE, etc.
+        mode: 'cors', // no-cors, cors, *same-origin
+        cache: 'no-cache', // *default, no-cache, reload, force-cache, only-if-cached
+        credentials: 'same-origin', // include, *same-origin, omit
+        headers: {
+            'Content-Type': 'application/json',
+            // 'Content-Type': 'application/x-www-form-urlencoded',
+        },
+        redirect: 'follow', // manual, *follow, error
+        referrer: 'no-referrer', // no-referrer, *client
+        body: JSON.stringify(data), // body data type must match "Content-Type" header
+    })
+    .then(response => response.json()); // parses JSON response into native JavaScript objects
+}
+
+/*item 추가 함수 */
+export var addItem = function (event) {
+    event.preventDefault();
+    const $app = document.querySelector('#app');
+    const $appenderForm = $app.querySelector('form[name="appenderForm"]');
+    const content = $appenderForm.querySelector('input').value.trim(); //앞 뒤 빈칸 제거
+    
+    if(content.length === 0) {
+      return alert("enter content");
+    }
+    
+    state.items.push({
+      id: state.items.length,
+      content: $appenderForm.querySelector('input').value,
+      isComplete: false,
+      createtime: Date.now(),
+    })  
+
+    postData('/app/items', {content: content})
+    .then(data => console.log(JSON.stringify(data))) // JSON-string from `response.json()` call
+    .catch(error => console.error(error));
+    
+    render();
+  }
+  
+  /*item 수정버튼 눌렀을 때 처리 함수 */
+export var editItem = function(event) {
+    state.idx = Number(event.target.dataset.key);
+    render();
+  }
+  
+  /* 수정 내용 반영 함수 */
+export var updateItem = function(event) {
+    event.preventDefault();
+    const content = event.target.querySelector('input').value.trim();
+  
+    if(content.length === 0) {
+      return alert("enter content");
+    }
+  
+    state.items[state.idx].content = content;
+
+    fetch(`/app/items/${state.idx +1}`, {
+      method: 'PUT', // *GET, POST, PUT, DELETE, etc.
+      mode: 'cors', // no-cors, cors, *same-origin
+      cache: 'no-cache', // *default, no-cache, reload, force-cache, only-if-cached
+      credentials: 'same-origin', // include, *same-origin, omit
+      headers: {
+          'Content-Type': 'application/json',
+          // 'Content-Type': 'application/x-www-form-urlencoded',
+      },
+      redirect: 'follow', // manual, *follow, error
+      referrer: 'no-referrer', // no-referrer, *client
+      body: JSON.stringify({content:content}), // body data type must match "Content-Type" header
+    }).then(res=>res.json())
+      .then(console.log("updated completed!"));
```

Comment:
```suggestion
      .then(() => console.log("updated completed!"));
```
이렇게 함수 내부에서 표현해줘야 돼요!
지금은 fetch가 완료되기 전에 console.log를 실행한답니다.